### ISSUE TITLE
Memory usage on Android can grow with repeated attaching of a long-lived Flutter engine to Flutter views

### ISSUE THREAD

**Issue by xster** [flutter/flutter#108601]
Making P1 to match b/237412494. 

Using https://github.com/flutter/samples/tree/main/add_to_app/books as an example, while repeatedly opening and closing each Flutter view (in that sample's case, click "Edit" to go to a Flutter details page, then popping back to the Android list) causes the "Others" memory category to grow by 2.1MB and "Native" category to grow by 15.1MB.

<img width="1975" alt="Screen Shot 2022-07-29 at 12 35 18 AM" src="https://user-images.githubusercontent.com/156888/181708627-f5c68f2a-347f-4b99-82f4-8b4d94c882c2.png">
 
In the "Others" category, I was able to observe the heap snapshot grow by 0.9MB
![9pGTpawoUHq5iGn](https://user-images.githubusercontent.com/156888/181708687-c7839ea2-e4fa-4ab6-b97c-4459be8692c5.png)

![4CA58XVXXzpEnDN](https://user-images.githubusercontent.com/156888/181708743-c3532745-73b6-4336-a6fc-dfc1cb816d17.png)

The main difference is the leftover image cache. Though that's already a known issue based on b/237630052.

**Comment by Hixie** [comment]
Weird that the "Graphics" memory is going down over the same period, are we just assigning it differently over time somehow? The total amount of usage seems stable on that chart.

**Comment by zanderso** [comment]
Internal P1 issues don't necessarily map to a Flutter P1 issue. I am dropping this to P3 to match the Flutter policy. cc @dnfield

**Comment by dnfield** [comment]
@xster - when I try to run books locally it crashes as soon as I try to tap an edit button. I've tried on a Pixel 6 pro and on a Moto G4, both running late versions of Android.

It also looks like your patch to help reproduce this more clearly hasn't landed in samples yet?

**Comment by dnfield** [comment]
Specifically https://github.com/flutter/samples/pull/1353

**Comment by xster** [comment]
> The total amount of usage seems stable on that chart.

I'm not sure that's the case. If you look at the trough of each open/close cycle, memory is creeping slightly. 

> Internal P1 issues don't necessarily map to a Flutter P1 issue.

I'm just following https://github.com/flutter/flutter/wiki/Triage#priority-labels `In general, P1 issues are bugs blocking top-tier customers from shipping immediately` which is generally the case. 

> it crashes as soon as I try to tap an edit button

are there logs in logcat? Can you file a separate bug? If our samples no longer run, we should definitely fix them.

**Comment by zanderso** [comment]
Thanks for clarifying that this is a blocker. Adding back P1 and assigning @dnfield.

**Comment by xster** [comment]
Moving to P2. I mistranslated from go/flutter-bugs#github-priority

**Comment by chinmaygarde** [comment]
@dnfield is OOO but from the last update, he could not reproduce this because the reproduction in question kept crashing.

**Comment by chinmaygarde** [comment]
@dnfield and @jason-simmons could not run the application at all. We need a runnable reproducible test case to make progress.

**Comment by zanderso** [comment]
@dnfield Could you describe the failure you're seeing? cc @xster

**Comment by jason-simmons** [comment]
When I tap the edit button in the Books application I get an exception.  It looks like it's coming from the Pigeon serializers:

```
08-22 13:06:15.747 10436 10436 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object java.util.Map.get(java.lang.Object)' on a null object reference
08-22 13:06:15.747 10436 10436 E AndroidRuntime: 	at dev.flutter.example.books.Api$Thumbnail.fromMap(Api.java:106)
08-22 13:06:15.747 10436 10436 E AndroidRuntime: 	at dev.flutter.example.books.Api$Book.fromMap(Api.java:88)
08-22 13:06:15.747 10436 10436 E AndroidRuntime: 	at dev.flutter.example.books.FlutterBookActivity.configureFlutterEngine(FlutterBookActivity.kt:71)
08-22 13:06:15.747 10436 10436 E AndroidRuntime: 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onAttach(FlutterActivityAndFragmentDelegate.java:206)
```

I tried regenerating the Pigeon code using the instructions in https://github.com/flutter/samples/tree/main/add_to_app/books.  But I was unable to get the resulting output to work.

**Comment by zanderso** [comment]
Maybe @gaaclarke can assist?

**Comment by dnfield** [comment]
You need to pull in @xster's latest PR (flutter/samples#1353). That sample is currently in a broken state.

**Comment by gaaclarke** [comment]
> Maybe @gaaclarke can assist?

Sounds like xiao's PR fixes this?  (https://github.com/flutter/samples/pull/1353)  Let me know if you need more eyes on this.  Xiao knows the most about it but I'm happy to look at it too.

**Comment by dnfield** [comment]
Per my last chat with @xster -

- we have a working repro of this now.
- The leak is relatively small and it's not clear that it scales up
- This deserves attention but is not currently blocking any internal customer launches. I'm lowering priority to P3 to get it off the critical issue triage radar.

**Comment by xster** [comment]
To clarify, internal user states that it is a launch blocker

**Comment by zanderso** [comment]
Bumping back to P2.

**Comment by dnfield** [comment]
@xster - when I run the sample app, I do see some native memory growth, but I also see ~18-20 megs of uncollected garbage after dismissing the Flutter UI. 

For example, in the run I'm looking at now I did back and forth about 4 times.

I see about 3 extra megs of graphics memory, which is probably the images in the cache.

I see 18.7 megs of uncollected garbage. I have 92.6 native memory used after, less uncollected garbage gives me 73.9, which would actually be slight less than the 74.5 megs used before. 

I can't seem to force a GC to run (at least, not in a way that makes a new heap snapshot show no remaining uncollected garbage).

Anyway, I think our first best path forward here i sto have some way to try to get the VM to more aggressively collect garbage in this case. I think you were talking with @a-siva about this recently.

I'm also a bit concerned about whether this issue truly mirrors the internal customer issue, as we're takling about what seems to be a one time fixed accumulation rather than any kind of ongoing build-up of memory after repeatedly showing and dismissing the UI. Happy to talk more about this, but that was part of why I had lowered the priority of this issue.

**Comment by xster** [comment]
Just to clarify my understanding, I was seeing a growth in both the "other" category and the "native" category. I was under the impression that VM held heap space that can be garbage collected all fit in the "other" category. Are you saying that both "other" and "native" would be affected by a GC? @a-siva could we test this theory? By having some custom engine brute force a GC and see if we can stop all growth?

**Comment by a-siva** [comment]
> Just to clarify my understanding, I was seeing a growth in both the "other" category and the "native" category. I was under the impression that VM held heap space that can be garbage collected all fit in the "other" category. Are you saying that both "other" and "native" would be affected by a GC? @a-siva could we test this theory? By having some custom engine brute force a GC and see if we can stop all growth?

Not sure I understand the comment about both "other" and "native" being affected by a GC, the recent engine change of not reporting huge external sizes for native objects and explicit disposal of these objects should ensure that GC will not in anyway effect the "native" footprint if it is coming from these native objects allocated by the engine.
I can try a local build of the engine with some quick hack to aggressively collect garbage on this example to see if there is any change, what would be the trigger point in the engine for when the Flutter UI is dismissed ?

**Comment by chinmaygarde** [comment]
| what would be the trigger point in the engine for when the Flutter UI is dismissed ?

Perhaps `PlatformView::OnPlatformViewDestroyed`?

**Comment by zanderso** [comment]
cc @iskakaushik

**Comment by xster** [comment]
Very excited about the GC mode work @iskakaushik. Just to clarify, I think there are 2 issues here. The Dart VM GC issue which a final GC during detach would solve (also see b/240654184), and there's a leak in "native". The latter is a separate issue.

**Comment by chinmaygarde** [comment]
From my understanding of the state of things, we think one leak is the pending collection and @iskakaushik s work on RAIL would partially address that. And then there is this other unexplained native memory growth that need to investigated. @xster Is that right? Are the two of the same priority or can be track them separately. Trying to figure out if we are overloading this issue.

**Comment by xster** [comment]
This bug is just for the native memory growth from repeated launch and dismiss. Let's keep the last VM GC with RAILs issue separate in b/240654184

**Comment by chinmaygarde** [comment]
From Triage: We are trying to find the right assignee for this. @iskakaushik AI was to help @a-siva with the RAIL hook in the engine. That work has been done and he as moved on. I don't think @dnfield is working on this either (though he did take a stab at it).

@xster @dnfield @a-siva @zanderso  Can we have a chat about this in person? We are confused in triage about what the remaining action items are.

**Comment by chinmaygarde** [comment]
From Triage: Per @a-siva from the Dart Sync, the team is going to take a look at this issue this week,

**Comment by xster** [comment]
To clarify, we should keep the GC related discussions separate. This bug is purely for the growth in the native C++ allocated space over time with repeated opens. 

I can reproduce this on head with https://github.com/flutter/samples/tree/main/add_to_app/books. 

After first launch and dismiss:
<img width="1453" alt="Screen Shot 2022-10-05 at 6 19 33 PM" src="https://user-images.githubusercontent.com/156888/194192726-642a1637-013b-4972-be43-53f32781e09b.png">

After a few launch and dismisses and a final Java GC:
<img width="1464" alt="Screen Shot 2022-10-05 at 6 19 45 PM" src="https://user-images.githubusercontent.com/156888/194192767-18079dc8-245e-4d36-84aa-e0831dfb2903.png">

**Comment by xster** [comment]
I didn't build a local engine to de-symbolicate, but there's a rough display of the "native" allocation growth:

I'm starting and closing Flutter one time and only capturing newly allocated and not deallocated memory since
<img width="1449" alt="Screen Shot 2022-10-05 at 6 27 40 PM" src="https://user-images.githubusercontent.com/156888/194193553-a14203ca-24a6-4d80-ab36-7a1e7a4c84a5.png">

The remaining leftovers from repeated launches is 11mb. 
<img width="2209" alt="Screen Shot 2022-10-05 at 6 26 36 PM" src="https://user-images.githubusercontent.com/156888/194193601-38159fcb-4ab4-4485-b82e-833072e06176.png">

Seems like half is from the embedding and half from somewhere deeper in the libapp.so.

**Comment by a-siva** [comment]
@rmacnak-google added a call in Dart C API called Dart_NotifyDetach here https://dart-review.googlesource.com/c/sdk/+/262520
He wanted to rename this method to match with Flutter's name in PlatformView::OnPlatformViewDestroyed.
Once the rename CL lands on the Dart side we will add the hook to make this call in PlatformView::OnPlatformViewDestroyed

**Comment by xster** [comment]
To disambiguate again, this issue is just for tracking native memory growth. Let's track the last GC issue separately in b/240654184

**Comment by zanderso** [comment]
From issue triage: The non-Dart heap component will be easier to investigate after b/240654184 is implemented.

**Comment by chinmaygarde** [comment]
From Triage: Per @dnfield investigation, it wasn't able to track down any appreciable instances of uncollected native memory and the majority of the allocations we related to the uncollected garbage. If the fix for that is separate and the uncollected native memory growth is small, is this still a P2?

**Comment by zanderso** [comment]
I think there are some crossed wires in this issue. My understanding is that there are two sources of retained memory across add-to-app instances.
1. Dart heap memory after Flutter add-to-app screen is dismissed. This is expected to be solved by adding a `Dart_NotifyDetach` call in the right place in the engine. Our understanding is that for the internal customer this is the bulk of the retained memory, and is the cause for this issue to be at P2.
2. A small (~10MB) amount of native heap memory accumulates across several add-to-app screens. The cause of this memory being retained is unknown. It sounds like the team is skeptical that this small amount of retained memory is really launch-blocking, and so the native-memory part of this issue should have a lower priority.

Do I have that right?

**Comment by a-siva** [comment]
Work related to cleaning up the Dart heap memory after Flutter add-to-app screen is dismissed has been completed with the new Dart_NotifyDestroyed api call and engine code being wired to call it.

**Comment by zanderso** [comment]
From triage: Since https://github.com/flutter/engine/pull/37539 is landed, this is ready for verification and/or further investigation. Routing initially to @dnfield for a quick check on that.

**Comment by chinmaygarde** [comment]
Removing Dans assignment since he is OOO. Lowering the priority as the linked patch has landed. We can close this after verification.

**Comment by zanderso** [comment]
My understanding is that the customer's issue has been addressed, and that no further follow-up is needed here, so I'm going to close this, but feel free to re-open if I'm mistaken.
